File: companies/appf/clock_walk.py
import unittest

class Solution:
    SIZE = 12

    def shortestDist(self, start, finish):
        clockwise = self._clockwiseDist(start, finish)
        counterClockwise = self._counterClockwiseDist(start, finish)

        if clockwise < counterClockwise:
            return ('clockwise', clockwise)
        return ('counter clockwise', counterClockwise)

    def _clockwiseDist(self, start, finish):
        return (self.SIZE + finish - start) % self.SIZE

    def _counterClockwiseDist(self, start, finish):
        return self.SIZE - self._clockwiseDist(start, finish)


class SolutionTest(unittest.TestCase):
    def testShortestDist(self):
        sol = Solution()

        self.assertEqual(('clockwise', 0), sol.shortestDist(1, 1))
        self.assertEqual(('clockwise', 5), sol.shortestDist(2, 7))
        self.assertEqual(('counter clockwise', 5), sol.shortestDist(7, 2))
        self.assertEqual(('counter clockwise', 4), sol.shortestDist(2, 10))
        self.assertEqual(('clockwise', 4), sol.shortestDist(10, 2))
        self.assertEqual(('clockwise', 3), sol.shortestDist(12, 3))

if __name__=="__main__":
    unittest.main()

File: companies/appf/clock_walk.rb
# Given two clock number a, b
# return the direction clock-wise or counter-clock-wise
# which give the fastest way to go from a to b

require "minitest/autorun"

class Solution
  SIZE = 12

  def shortest_dist(start, fin)
    clock_wise = clockwise_dist(start, fin)
    counter_clockwise = counter_clockwise_dist(start, fin)

    if clock_wise < counter_clockwise
      ['clockwise', clock_wise]
    else
      ['counter clockwise', counter_clockwise]
    end
  end

  private

  def clockwise_dist(start, fin)
    (SIZE + fin - start) % SIZE
  end

  def counter_clockwise_dist(start, fin)
    SIZE - clockwise_dist(start, fin)
  end
end

class SolutionTest < Minitest::Test
  def test_shortest_dist
    sol = Solution.new
    assert_equal ['clockwise', 0], sol.shortest_dist(1, 1)
    assert_equal ['clockwise', 5], sol.shortest_dist(2, 7)
    assert_equal ['counter clockwise', 5], sol.shortest_dist(7, 2)
    assert_equal ['counter clockwise', 4], sol.shortest_dist(2, 10)
    assert_equal ['clockwise', 4], sol.shortest_dist(10, 2)
    assert_equal ['clockwise', 3], sol.shortest_dist(12, 3)
  end
end
